extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -665,180 +694,185 @@ public String getPosition(StreamPartition<String> partition)
-        return null;
-      }
-      try {
-        // we call getRecords with limit 1000 to make sure that we can find the first (earliest) record in the shard.
-        // In the case where the shard is constantly removing records that are past their retention period, it is possible
-        // that we never find the first record in the shard if we use a limit of 1.
-        recordsResult = kinesis.getRecords(new GetRecordsRequest().withShardIterator(shardIterator).withLimit(1000));
-      }
-      catch (ProvisionedThroughputExceededException e) {
-        log.warn(
-            e,
-            "encountered ProvisionedThroughputExceededException while fetching records, this means "
-            + "that the request rate for the stream is too high, or the requested data is too large for "
-            + "the available throughput. Reduce the frequency or size of your requests. Consider increasing "
-            + "the number of shards to increase throughput."
-        );
+    return wrapExceptions(() -> {
+      String shardIterator =
+          kinesis.getShardIterator(partition.getStream(), partition.getPartitionId(), iteratorEnum.toString())
+                 .getShardIterator();
+      long timeoutMillis = System.currentTimeMillis() + fetchSequenceNumberTimeout;
+      GetRecordsResult recordsResult = null;
+
+      while (shardIterator != null && System.currentTimeMillis() < timeoutMillis) {
+
+        if (closed) {
+          log.info("KinesisRecordSupplier closed while fetching sequenceNumber");
+          return null;
+        }
         try {
-          Thread.sleep(PROVISIONED_THROUGHPUT_EXCEEDED_BACKOFF_MS);
-          continue;
+          // we call getRecords with limit 1000 to make sure that we can find the first (earliest) record in the shard.
+          // In the case where the shard is constantly removing records that are past their retention period, it is possible
+          // that we never find the first record in the shard if we use a limit of 1.
+          recordsResult = kinesis.getRecords(
+              new GetRecordsRequest().withShardIterator(shardIterator).withLimit(FETCH_SEQUENCE_NUMBER_RECORD_COUNT)
+          );
         }
-        catch (InterruptedException e1) {
-          log.warn(e1, "Thread interrupted!");
-          Thread.currentThread().interrupt();
-          break;
+        catch (ProvisionedThroughputExceededException e) {
+          log.warn(
+              e,
+              "encountered ProvisionedThroughputExceededException while fetching records, this means "
+              + "that the request rate for the stream is too high, or the requested data is too large for "
+              + "the available throughput. Reduce the frequency or size of your requests. Consider increasing "
+              + "the number of shards to increase throughput."
+          );
+          try {
+            Thread.sleep(PROVISIONED_THROUGHPUT_EXCEEDED_BACKOFF_MS);
```

Comment:
It seems better to me, and more consistent with the rest of the codebase, though I'm not sure what to do for retry count since the current behavior would be an indefinite number of retries in the case of `ProvisionedThroughputExceededException`. However, I picked 10, though I can do higher if we are worried this will commonly happen...

Reply:
I guess we can add another mode to `RetryUtils.retry()` that checks the timeout instead of retryCount, but the current way looks ok to me.